sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -92,6 +92,11 @@ case object REPARTITION extends ShuffleOrigin
 // a certain partition number. Spark can't optimize it.
 case object REPARTITION_WITH_NUM extends ShuffleOrigin
 
+// Indicates that the shuffle operator was added by the user-specified repartition operator.
+// Spark first tries to coalesce partitions, if it cannot be coalesced, then try to use the
+// local shuffle reader.
+case object REPARTITION_WITHOUT_COL_AND_NUM extends ShuffleOrigin
```

Comment:
We can thing about the naming a bit more. e.g. `REPARTITION_WITH_COL`, `REPARTITION_WITH_NUM`, `REPARTITION_WITH_NOTHING`

Reply:
cc @yaooqinn @JkSelf

Reply:
Are we use `_WITH_XXX` to indicate the point from which shuffle starts?

Reply:
yea, it describes the user intention and indicates how spark can optimize it without breaking user intention

Reply:
How about `REPARTITION_BY_COL`, `REPARTITION_BY_NUM`, `REPARTITION`

Reply:
or `REPARTITION` -> `REPARTITION_BY_NONE`

Reply:
SGTM. +1 for REPARTITION_BY_NONE

Reply:
Here we have 4 cases for repartition: only num, only col, both num and col, neither num or col. Why we don't handle the both num and col case?

Reply:
See https://github.com/apache/spark/pull/32781#discussion_r647998499